### DISCUSSION TITLE
Method toJSON does not renders  all fetched data

### DISCUSSION THREAD

**Discussion by triadium** [mikro-orm/mikro-orm#6081] [General]
### Describe the bug

We use the method `toJSON` to return data from the service, but the client side began to fall with an error and it was discovered that not all nested data are correctly output from the fetched data. We decided to switch to the method `toPOJO`, but it has its own peculiarities.

### Reproduction

```ts
const MikroOrmConfig: Options = {
  dbName: configService.get('MODYNO_POSTGRES_DB'),
  driver: PostgreSqlDriver,
  namingStrategy: EntityCaseNamingStrategy,
  entities: [],
  host: configService.get('MODYNO_POSTGRES_HOST'),
  password: configService.get('MODYNO_POSTGRES_PASSWORD'),
  port: configService.get('MODYNO_POSTGRES_PORT'),
  user: configService.get('MODYNO_POSTGRES_USER'),
  metadataCache: {
    enabled: false
  },
  debug: true,
}

const ASchema = new EntitySchema<any>({
  name: 'A',
  properties: {
    id: { type: 'uuid', defaultRaw: 'gen_random_uuid()', primary: true },
    aname: { type: 'text', nullable: true },
    refB: { kind: 'm:1', entity: 'B', nullable: true, deleteRule: 'cascade', updateRule: 'no action' },
  },
});
const BSchema = new EntitySchema<any>({
  name: 'B',
  properties: {
    id: { type: 'uuid', defaultRaw: 'gen_random_uuid()', primary: true },
    bname: { type: 'text', nullable: true },
  },
});
const DSchema = new EntitySchema<any>({
  name: 'D',
  schema: 'xab_ru',
  properties: {
    id: { type: 'uuid', defaultRaw: 'gen_random_uuid()', primary: true },
    dname: { type: 'text', nullable: true },
    refA: { kind: 'm:1', entity: 'A', nullable: true, deleteRule: 'cascade', updateRule: 'no action' },
  },
});

MikroOrmConfig.entities.push(ASchema);
MikroOrmConfig.entities.push(BSchema);
MikroOrmConfig.entities.push(DSchema);

const orm = await MikroORM.init(MikroOrmConfig)
const fem = orm.em.fork() as EntityManager

const objB = fem.create(BSchema, { bname: 'B-Name' });
const objA = fem.create(ASchema, { aname: 'A-Name', refB: objB });
const objD = fem.create(DSchema, { dname: 'D-Name', refA: objA });

await fem.flush();

const multiD = await fem.find(DSchema, {}, 
  { 
    fields: ['*', 'refA.*', 'refA.refB.*'], 
    strategy: 'joined', 
    orderBy: { refA: { label: QueryOrder.DESC_NULLS_LAST } }, 
    populate: ['refA.refB.*'] 
});
console.log(`%o`, multiD[0]);
console.log(`%o`, multiD[0].toJSON());
console.log(`%o`, wrap(multiD[0]).toPOJO());
```
Output:

```
-- Direct log
D {
  dname: 'D-Name',  
  refA: A {    
    aname: 'A-Name',    
    refB: B {      
      bname: 'B-Name',
      id: 'b22425c6-00e8-43b8-ac66-b3eb88bab1ca'
    },
    id: '6b0c6d6a-7aa1-4ab6-af46-9c86328cb9ff'
  },
  id: '5308d847-3823-4894-aa72-442ba36b58a9'
}
-- toJSON
{
  id: '5308d847-3823-4894-aa72-442ba36b58a9',
  dname: 'D-Name',  
  refA: {
    id: '6b0c6d6a-7aa1-4ab6-af46-9c86328cb9ff',    
    aname: 'A-Name',
    refB: 'b22425c6-00e8-43b8-ac66-b3eb88bab1ca'    
  }  
}
--wrap.toPOJO
{
  id: '5308d847-3823-4894-aa72-442ba36b58a9',  
  dname: 'D-Name',
  refA: {
    id: '6b0c6d6a-7aa1-4ab6-af46-9c86328cb9ff',    
    aname: 'A-Name',
    refB: {
      id: 'b22425c6-00e8-43b8-ac66-b3eb88bab1ca',      
      bname: 'B-Name'
    }
  }
}
```
 

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.3.11

### Node.js version

v20.11.1

### Operating system

Windows 11

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
Please provide a complete reproduction, the one you provided is not doing what you say. And please use this format instead:

https://github.com/mikro-orm/reproduction

Here is yours converted to this format already, and it works just fine:

```ts
import { EntitySchema, MikroORM } from '@mikro-orm/sqlite';

const ASchema = new EntitySchema({
  name: 'A',
  properties: {
    id: { type: 'number', primary: true },
    aname: { type: 'text', nullable: true },
    refB: { kind: 'm:1', entity: 'B', nullable: true },
  },
});
const BSchema = new EntitySchema({
  name: 'B',
  properties: {
    id: { type: 'number', primary: true },
    bname: { type: 'text', nullable: true },
  },
});
const DSchema = new EntitySchema({
  name: 'D',
  properties: {
    id: { type: 'number', primary: true },
    dname: { type: 'text', nullable: true },
    refA: { kind: 'm:1', entity: 'A', nullable: true },
  },
});

let orm: MikroORM;

beforeAll(async () => {
  orm = await MikroORM.init({
    entities: [ASchema, BSchema, DSchema],
    debug: true,
    dbName: ':memory:',
  });
  await orm.schema.createSchema();

  const objB = orm.em.create(BSchema, { bname: 'B-Name' });
  const objA = orm.em.create(ASchema, { aname: 'A-Name', refB: objB });
  const objD = orm.em.create(DSchema, { dname: 'D-Name', refA: objA });
  await orm.em.flush();
  orm.em.clear();
});

afterAll(() => orm.close(true));

test('6080', async () => {
  const multiD = await orm.em.find(DSchema, {},
    {
      fields: ['*', 'refA.*', 'refA.refB.*'],
      populate: ['refA.refB.*'],
    });
  console.log(`%o`, multiD[0]);
  console.log(`%o`, multiD[0].toJSON());
});
```

>       populate: ['refA.refB.*'],

Note that this is quite weird on its own, what you want from that asterisk? There are no relations on B.

**Reply by triadium** [reply]
Sorry for the late response, but I found why your code does not give an error, but mine did. When I copied the code to issue, it was cut off and copied less than it originally was. The problem is in multiple requests with a single fork of the manager.

```ts
import { EntitySchema, MikroORM } from '@mikro-orm/sqlite';

const ASchema = new EntitySchema({
  name: 'A',
  properties: {
    id: { type: 'number', primary: true },
    aname: { type: 'text', nullable: true },
    refB: { kind: 'm:1', entity: 'B', nullable: true },
  },
});
const BSchema = new EntitySchema({
  name: 'B',
  properties: {
    id: { type: 'number', primary: true },
    bname: { type: 'text', nullable: true },
  },
});
const DSchema = new EntitySchema({
  name: 'D',
  properties: {
    id: { type: 'number', primary: true },
    dname: { type: 'text', nullable: true },
    refA: { kind: 'm:1', entity: 'A', nullable: true },
  },
});

let orm: MikroORM;
let oneDId: number;

beforeAll(async () => {
  orm = await MikroORM.init({
    entities: [ASchema, BSchema, DSchema],
    debug: true,
    dbName: ':memory:',
  });
  await orm.schema.createSchema();

  const fem = orm.em.fork()

  const objB = fem.create(BSchema, { bname: 'B-Name' });
  const objA = fem.create(ASchema, { aname: 'A-Name', refB: objB });
  const objD = fem.create(DSchema, { dname: 'D-Name', refA: objA });
  await fem.flush();
  oneDId = objD.id;
  fem.clear();
});

afterAll(() => orm.close(true));

test('6080', async () => {
  const fem = orm.em.fork()
  const oneD = await fem.findOneOrFail(DSchema, { id: oneDId },
    {
      fields: ['*', 'refA.*'/*, no need here 'refA.refB.*' yet */],
    });
  console.log(`%o`, oneD.toJSON());

  // ... some large code without fem.clear()

  // fem.clear() helps to resolve the problem
  const multiD = await fem.find(DSchema, {},
    {
      fields: ['*', 'refA.*', 'refA.refB.*'],
    });
  console.log(`%o`, multiD[0].toJSON());
});
```

**Reply by B4nan** [reply]
>  multiple requests with a single fork of the manager.

You shouldn't do that, not just for this reason. Every request should have its own context.

https://mikro-orm.io/docs/identity-map#why-is-request-context-needed

**Reply by triadium** [reply]
> The problem is in multiple requests with a single fork of the manager.

I didn't mean requests to the application, but to the database, i.e. queries.  Sorry, my typo.

Our case looks like this: 

within the framework of a single request to the application, one "central" object is selected. 

Then a selection is made by references through an intermediate table, and list of other objects is selected according to certain filters. 

This list - "other objects" - may contain the "central" object itself (this is acceptable for us). 

And then everything breaks down, because we need to enrich the objects with additional data, but this "central" object turns out to be the same as in the first selection. I thought that the system would automatically refresh it. 

Now I'm em cleaning before "find" so that everything works correctly, but this behavior is not obvious.

**Reply by triadium** [reply]
I found the `disableIdentityMap` parameter and saw that "fork.clear()" is called in the code. 

I need to know - when an object is not refreshed in IdentityMap after the first query by the second query with extended fields, it is by design or an error?

**Reply by B4nan** [reply]
refreshing is always opt-in. if you query an object by its PK, there is no query, if you query by something else, then a query is made and the ORM merges the data automatically, skipping those that you already changed in memory.

**Reply by B4nan** [reply]
i will have a closer look at your repro, i think i know where the problem is, and i don't think its about refreshing, it sounds more like a problem with merging the loading hints (`populate` and `fields`).

you could also try the new `wrap().setSerializationContext()` which was added to allow explicit control over the hints that are used during serialization. this should be effectively the same if the data is already there:

```ts
const multiD = await fem.find(DSchema, {}, {
  fields: ['*', 'refA.*', 'refA.refB.*'],
});
multiD.forEach(d => wrap(d).setSerializationContext({
  fields: ['*', 'refA.*', 'refA.refB.*'],
});
console.log(`%o`, multiD[0].toJSON());
```

**Reply by triadium** [reply]
It doesn't seem to work
```ts
multiD.forEach(d => wrap(d).setSerializationContext({
  fields: ['*', 'refA.*', 'refA.refB.*'],
});
```

If I fix the internal isPopulated function so that it returns true, then everything works fine, but I understand that this is not how it should be solved.

[mikro-orm-reproduction/node_modules/@mikro-orm/core/serialization/EntityTransformer.js](https://github.com/mikro-orm/mikro-orm/blob/9e576479585886bb7d983d82509ae969584ef7cd/packages/core/src/serialization/EntityTransformer.ts#L213)
